### PR TITLE
Convert cmd/syncer/main.go to be a cobra.Command

### LINKED ISSUE #553: Convert cmd/syncer/main.go to be a cobra.Command

### PR DESCRIPTION
**Pull request by RAMESSESII2**
### Migrating to cobra.Command
* added cobra command equivalent of existing go flags.
* moved the main logic from ./cmd/syncer/main.go to ./cmd/syncer/cmd/syncer.go
* created a new exported ```Options struct``` for flags binding.
* syncedResourceTypes is a flag now
image

## Testing
* Build the source code
```bash
go build ./cmd/syncer
```
* Run it
```bash
./syncer -h
```
- [x] Fixes #553

### CODE REVIEW

File: cmd/syncer/main.go [hunk 1]
```diff
@@ -16,88 +16,6 @@ limitations under the License.
...
-
-	klog.Infoln("Starting workers")
-	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), *fromClusterName, *pclusterID, numThreads); err != nil {
-		klog.Fatal(err)
-	}
-
-	<-ctx.Done()
-
-	klog.Infoln("Stopping workers")
+	ExecuteCmd()
```
Comment (ncdc) [inline cmd/syncer/main.go:20]: I'd look into following a pattern like you see at https://github.com/kubernetes/kubernetes/blob/88f97283393715c7fa870dd90be0fd0d0faad6bd/cmd/kube-controller-manager/controller-manager.go:

```go
func main() {
	command := app.NewControllerManagerCommand()
	code := cli.Run(command)
	os.Exit(code)
}
```

The reason to put the cobra command in a separate package is so others can import it & use it, and we can test it. The example I shared above puts it in `app`, which would be ok, or you could put it in `cmd` (`cmd/syncer/cmd`).

File: cmd/syncer/root.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2021 The KCP Authors.
```
Comment (ncdc) [inline cmd/syncer/root.go:2]:
```suggestion
Copyright 2022 The KCP Authors.
```

File: cmd/syncer/root.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
+/*
```
Comment (ncdc) [inline cmd/syncer/root.go:1]: Please move this file into a subpackage (see comment above)

File: cmd/syncer/root.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+const numThreads = 2
+
+var (
+	fromKubeconfig      string
+	fromClusterName     string
+	toKubeconfig        string
+	toContext           string
+	pclusterID          string
+	syncedResourceTypes []string
+)
```
Comment (ncdc) [inline cmd/syncer/root.go:45]: Please move this into an `Options` struct - see https://github.com/kcp-dev/kcp/blob/ba104a795fa57e7f452a91b5fb30f96463aa1599/cmd/ingress-controller/main.go#L122-L150 for a good reference example.

File: cmd/syncer/root.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+var rootCmd = &cobra.Command{
```
Comment (ncdc) [inline cmd/syncer/root.go:47]: Instead of making this a private package variable, let's create a constructor (`NewSyncerCommand`) that returns the `*cobra.Command`. This function can also set up the options & bind flags.

File: cmd/syncer/root.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+		synchronise()
+	},
+}
+
+func synchronise() {
+
+	// Create a client to dynamically watch "from".
+	if fromClusterName == "" {
+		klog.Fatal("--from_cluster is required")
```
Comment (ncdc) [inline cmd/syncer/root.go:66]: Let's move validation to `Options.Validate()`

File: cmd/syncer/root.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			klog.Fatal(err)
+		}
+	} else {
+		toConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGILL, syscall.SIGINT)
```
Comment (ncdc) [inline cmd/syncer/root.go:95]: Since you're in here, please do #552 😄

File: cmd/syncer/root.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	defer cancel()
+
+	klog.Infoln("Starting workers")
```
Comment (ncdc) [inline cmd/syncer/root.go:98]: Can remove this - code inside `syncer.StartSyncer` does similar logging (with more details)

File: cmd/syncer/root.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), fromClusterName, pclusterID, numThreads); err != nil {
+		klog.Fatal(err)
+	}
+
+	<-ctx.Done()
+
+	klog.Infoln("Stopping workers")
```
Comment (ncdc) [inline cmd/syncer/root.go:105]: Can remove this - code inside `syncer.StartSyncer` does similar logging (with more details)

File: cmd/syncer/root.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+	// Create a client to dynamically watch "from".
+	if fromClusterName == "" {
+		klog.Fatal("--from_cluster is required")
+	}
+
+	fromConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&clientcmd.ClientConfigLoadingRules{ExplicitPath: fromKubeconfig}, nil).ClientConfig()
+	if err != nil {
+		klog.Fatal(err)
```
Comment (ncdc) [inline cmd/syncer/root.go:72]: Let's return any `err` instead of doing `klog.Fatal`.

File: cmd/syncer/root.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	<-ctx.Done()
+
+	klog.Infoln("Stopping workers")
+}
+
+func ExecuteCmd() {
+	rootCmd.SetHelpFunc(Help)
```
Comment (ncdc) [inline cmd/syncer/root.go:109]: Do we need a custom help func?

File: cmd/syncer/root.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), fromClusterName, pclusterID, numThreads); err != nil {
+		klog.Fatal(err)
+	}
+
+	<-ctx.Done()
+
+	klog.Infoln("Stopping workers")
+}
+
+func ExecuteCmd() {
```
Comment (ncdc) [inline cmd/syncer/root.go:108]: Probably don't need this function any more, if you're doing
```go
code := cli.Run(command)
os.Exit(code)
```
in main.go

**Review by ncdc — CHANGES_REQUESTED** [review]
@RAMESSESII2 thanks so much for the PR! If you have any questions about my review comments, please let me know.

File: cmd/syncer/main.go (hunk 1 above)
Comment (RAMESSESII2) [reply cmd/syncer/main.go:20]: great idea, in fact, I already wanted to. I'm on it

File: cmd/syncer/root.go (hunk 2 above)
Comment (RAMESSESII2) [reply cmd/syncer/root.go:2]: done

File: cmd/syncer/root.go (hunk 3 above)
Comment (RAMESSESII2) [reply cmd/syncer/root.go:1]: done

File: cmd/syncer/main.go (hunk 1 above)
Comment (RAMESSESII2) [reply cmd/syncer/main.go:20]: done

File: cmd/syncer/root.go (hunk 4 above)
Comment (RAMESSESII2) [reply cmd/syncer/root.go:45]: @ncdc, should I keep the Logs field as well in the Syncer Options struct? Will there be any future prospects to it? WDYT

File: cmd/syncer/root.go (hunk 5 above)
Comment (RAMESSESII2) [reply cmd/syncer/root.go:47]: done

File: cmd/syncer/root.go (hunk 4 above)
Comment (ncdc) [reply cmd/syncer/root.go:45]: Yes, please. It configures `--v` and related flags.

File: cmd/syncer/root.go (hunk 11 above)
Comment (RAMESSESII2) [reply cmd/syncer/root.go:109]: Not really, it's just originally, we had ``` 	fmt.Fprintf(os.Stderr, "Usage of %s [resourceTypes...]:\n", path.Base(os.Args[0])) ``` - inside flag.Usage
although yeah we don't really required it, cobra help will be enough ig.

File: cmd/cluster-controller/main.go [hunk 13]
```diff
@@ -75,8 +74,7 @@ func (o *options) Validate() error {
 
 func main() {
 	// Setup signal handler for a cleaner shutdown
-	ctx, cancel := signal.NotifyContext(context.Background(), os.Kill, os.Interrupt)
-	defer cancel()
+	ctx := genericapiserver.SetupSignalContext()
```
Comment (ncdc) [inline cmd/cluster-controller/main.go:77]: Re signal handling - sorry - I meant to only do the change specific to syncer. But that's awesome that you've done the others! Would you mind splitting them into a separate PR (and leave the syncer changes in this one)?

File: cmd/syncer/cmd/root.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/component-base/logs"
+	"k8s.io/klog/v2"
+)
+
+const numThreads = 2
+
+func NewSyncerCommand() *cobra.Command {
+
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:38]:
```suggestion
```

File: cmd/syncer/cmd/root.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+	options := NewDefaultOptions()
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:39]:
```suggestion
	options := NewDefaultOptions()
	
```

File: cmd/syncer/cmd/root.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+	return nil
+}
+
+type Options struct {
+	fromKubeconfig  string
+	fromClusterName string
+	toKubeconfig    string
+	toContext       string
+	pclusterID      string
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:108]: Please make all of these fields exported - anyone programmatically creating & using an `Options` struct from outside this package needs to be able to set these values.

File: cmd/syncer/cmd/root.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+		pclusterID:      "",
+		Logs:            logs.NewOptions(),
+	}
+}
+
+func (options *Options) BindFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&options.fromKubeconfig, "from_kubeconfig", options.fromKubeconfig, "Kubeconfig file for -from cluster.")
+	fs.StringVar(&options.fromClusterName, "from_cluster", options.fromClusterName, "Name of the -from logical cluster.")
+	fs.StringVar(&options.toKubeconfig, "to_kubeconfig", options.toKubeconfig, "Kubeconfig file for -to cluster. If not set, the InCluster configuration will be used.")
+	fs.StringVar(&options.toContext, "to_context", options.toContext, "Context to use in the Kubeconfig file for -to cluster, instead of the current context.")
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:127]: Please convert the flag names to use dashes (`-`) to separate words instead of underscores (`_`)

File: cmd/syncer/main.go [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KCP Authors.
+Copyright 2022 The KCP Authors.
```
Comment (ncdc) [inline cmd/syncer/main.go:2]: Please don't change copyright dates of existing files (or if you're going to, I'd say only make additive changes; e.g. 2021-2022)

File: cmd/virtual-workspaces/main.go [hunk 19]
```diff
@@ -33,13 +33,13 @@ import (
 )
 
 func main() {
-	stopCh := genericapiserver.SetupSignalHandler()
+	ctx := genericapiserver.SetupSignalContext()
```
Comment (ncdc) [inline cmd/virtual-workspaces/main.go:36]: Split into separate PR (just focused on signal handling consistency)

File: cmd/syncer/cmd/root.go [hunk 20]
```diff
@@ -0,0 +1,131 @@
+/*
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:1]: nit: I would call this file syncer.go

File: cmd/syncer/cmd/root.go [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+	options := NewDefaultOptions()
+	syncerCommand := &cobra.Command{
+		Use:   "syncer",
+		Short: "Synchronizes resources in `kcp` assigned to the clusters",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := options.validate(args)
+			return err
+		},
+	}
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:47]:
```suggestion
	}
	
```

File: cmd/syncer/cmd/root.go [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+	options.BindFlags(syncerCommand.Flags())
+	return syncerCommand
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:49]:
```suggestion

	return syncerCommand
```

File: cmd/syncer/cmd/root.go [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func (options *Options) validate(args []string) error {
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:52]:
```suggestion
func (options *Options) Validate(args []string) error {
```

This needs to be exported so third party packages can invoke it

File: cmd/syncer/cmd/root.go [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+	//non-flag arguements
+	syncedResourceTypes := args
+	if len(syncedResourceTypes) == 0 {
+		syncedResourceTypes = []string{"deployments.apps"}
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:56]: I would move the defaulting logic into a new Complete() function.

File: cmd/syncer/cmd/root.go [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+			&toOverrides).ClientConfig()
+		if err != nil {
+			return err
+		}
+	} else {
+		toConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return err
+	}
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:90]: I think I would probably move this into Complete() too. Anything you're creating as variables in here could be stored as unexported fields in Options, then used in Run()

File: cmd/syncer/cmd/root.go [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+
+	ctx := genericapiserver.SetupSignalContext()
+
+	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), options.fromClusterName, options.pclusterID, numThreads); err != nil {
+		return err
+	}
+
+	<-ctx.Done()
```
Comment (ncdc) [inline cmd/syncer/cmd/root.go:98]: Please move this to a new function called Run()

**Review by ncdc — CHANGES_REQUESTED** [review]

File: cmd/syncer/cmd/root.go (hunk 24 above)
Comment (ncdc) [reply cmd/syncer/cmd/root.go:56]: cc @sttts would you agree with this comment + the ones below?

File: cmd/cluster-controller/main.go (hunk 13 above)
Comment (RAMESSESII2) [reply cmd/cluster-controller/main.go:77]: yeah sure.

File: cmd/syncer/cmd/root.go (hunk 16 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/root.go:108]: sorry, I missed that, nevertheless, DONE

File: cmd/syncer/cmd/root.go (hunk 23 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/root.go:52]: yeah, right 👍🏼

File: cmd/syncer/cmd/root.go (hunk 25 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/root.go:90]: yeah sure, that makes more sense.

**Comment by RAMESSESII2** [comment]
Hi @ncdc , I think I've incorporated the required changes. Please do check if I missed anything, thanks 😺

File: cmd/syncer/cmd/syncer.go [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+	return syncerCommand
+}
+
+func (options *Options) Validate() error {
+	err := options.Complete()
+	if err != nil {
+		return err
+	}
+
+	err = options.Run()
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:68]: this is odd. Why run in validation?

File: cmd/syncer/cmd/syncer.go [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	//bind flags
+	options.BindFlags(syncerCommand.Flags())
+
+	return syncerCommand
+}
+
+func (options *Options) Validate() error {
+	err := options.Complete()
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:63]: would not expect completion inside of validate. Validate should be side-effect free, otherwise it would be called ValidateAndDoX

File: cmd/syncer/cmd/syncer.go [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	return nil
+}
+
+func (options *Options) Complete() error {
+	if len(options.SyncedResourceTypes) == 0 {
+		options.SyncedResourceTypes = []string{"deployments.apps"}
+		klog.Infoln("No resource types provided; using defaults.")
+	}
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:80]: could we have that as default list in the Options struct on construction? Or is this a problem that a string set flag then cannot override it?

File: cmd/syncer/cmd/syncer.go [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+
+	return nil
+}
+
+type Options struct {
+	FromKubeconfig  string
+	FromClusterName string
+	ToKubeconfig    string
+	ToContext       string
+	PclusterID      string
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:134]: WorkloadClusterName

File: cmd/syncer/main.go [hunk 31]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KCP Authors.
+Copyright 2021-2022 The KCP Authors.
```
Comment (sttts) [inline cmd/syncer/main.go:2]: just 2021, no range

File: cmd/syncer/main.go [hunk 32]
```diff
@@ -17,87 +17,14 @@ limitations under the License.
...
-
-var (
-	fromKubeconfig  = flag.String("from_kubeconfig", "", "Kubeconfig file for -from cluster.")
-	fromClusterName = flag.String("from_cluster", "", "Name of the -from logical cluster.")
-	toKubeconfig    = flag.String("to_kubeconfig", "", "Kubeconfig file for -to cluster. If not set, the InCluster configuration will be used.")
-	toContext       = flag.String("to_context", "", "Context to use in the Kubeconfig file for -to cluster, instead of the current context.")
-	pclusterID      = flag.String("cluster", "",
-		fmt.Sprintf("ID of the -to cluster. Resources with this ID set in the '%s' label will be synced.", nscontroller.ClusterLabel))
+	"github.com/kcp-dev/kcp/cmd/syncer/cmd"
+	"k8s.io/component-base/cli"
```
Comment (sttts) [inline cmd/syncer/main.go:23]: we separate the kcp imports, i.e.
```
import (
   <go imports>

   <external imports>

   <kcp imports>
)
```
```

**Review by ncdc — CHANGES_REQUESTED** [review]
The changes are looking better! Thanks for your hard work!

File: cmd/syncer/cmd/syncer.go (hunk 27 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:68]: yeah I realise that doesn't make any sense.

File: cmd/syncer/cmd/syncer.go [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+const numThreads = 2
+
+func NewSyncerCommand() *cobra.Command {
+	options := NewDefaultOptions()
+
+	syncerCommand := &cobra.Command{
+		Use:   "syncer",
+		Short: "Synchronizes resources in `kcp` assigned to the clusters",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.SyncedResourceTypes = args
```
[the remark this answers is not in the document]
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:44]: Sure, agreed

File: cmd/syncer/cmd/syncer.go [hunk 34]
```diff
@@ -0,0 +1,164 @@
...
+func (options *Options) Complete() error {
+	if len(options.SyncedResourceTypes) == 0 {
+		options.SyncedResourceTypes = []string{"deployments.apps"}
+		klog.Infoln("No resource types provided; using defaults.")
+	}
+	klog.Infof("Syncing the following resource types: %s", options.SyncedResourceTypes)
+
+	// Create a client to dynamically watch "from".
+	if options.FromClusterName == "" {
+		return errors.New("--from_cluster is required")
```
[the remark this answers is not in the document]
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:85]: sorry, missed it

File: cmd/syncer/cmd/syncer.go (hunk 28 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:63]: sure, make sense

File: cmd/syncer/cmd/syncer.go (hunk 29 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:80]: Hey @ncdc , by "change", do you mean making ```SyncedResourceTypes``` a flag (StringArrayVar type)? I think that could be done.

File: cmd/syncer/cmd/syncer.go (hunk 30 above)
Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:134]: Hey @sttts, I'm sorry, I'm new to KCP but if I'm correct do you mean ClusterName is the same as WorkloadClusterName? And should I change ```FromClusterName``` to ```FromWorkloadClusterName```?

File: cmd/syncer/cmd/syncer.go (hunk 29 above)
Comment (ncdc) [reply cmd/syncer/cmd/syncer.go:80]: @RAMESSESII2 that's right, we could make SyncedResourceTypes a flag.

**Comment by RAMESSESII2** [comment]
Hey @ncdc, i think I've made the requested changes but still this can always be improved so let me know. 😄

**Comment by ncdc** [comment]
@marun would you be willing to take over reviewing this?

**Review by marun — APPROVED** [review]
I've manually tested this to run the syncer with the cluster e2e.

File: cmd/syncer/cmd/syncer.go [hunk 35]
```diff
@@ -0,0 +1,150 @@
...
+	FromClusterName     string
+	ToKubeconfig        string
+	ToContext           string
+	PclusterID          string
+	Logs                *logs.Options
+	SyncedResourceTypes []string
+
+	//non-exported members
+	toConfig   *rest.Config
+	fromConfig *rest.Config
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:125]: this does not belong into options.

File: cmd/syncer/cmd/syncer.go [hunk 36]
```diff
@@ -0,0 +1,150 @@
...
+
+func (options *Options) Run(ctx context.Context) error {
+	if err := syncer.StartSyncer(ctx, options.fromConfig, options.toConfig, sets.NewString(options.SyncedResourceTypes...), options.FromClusterName, options.PclusterID, numThreads); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type Options struct {
```
Comment (sttts) [inline cmd/syncer/cmd/syncer.go:114]: please move this into a sub-folder "options". Otherwise we cannot compose it.

File: cmd/syncer/cmd/syncer.go (hunk 35 above)
Comment (sttts) [reply cmd/syncer/cmd/syncer.go:125]: These are derived from the options, aren't they?

Usually we have only flags values (with very few exceptions) in option structs. Would put those into cmd/syncer/options, with AddFlags, Complete and Validate methods.

Then at the cmd/syncer/cmd level you will use the option structs from the options package and turn e.g. the kubeconfig path name into rest.Configs.

File: cmd/syncer/cmd/syncer.go (hunk 36 above)
Comment (marun) [reply cmd/syncer/cmd/syncer.go:114]: What are we intending to compose it with?

Comment (sttts) [reply cmd/syncer/cmd/syncer.go:114]: for example run in-process. We have it everywhere else. You won't know how to compose before you do it. Options struct are made exactly for that.

Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:114]: Hi @sttts, I agree, it makes sense to put Options struct in its own package. And then let's say I create new Struct which will eventually be composed of options and the derived variables(toConfig and fromConfig). Something like this -
```
type SyncerOptions struct{
         options *options.Options
         toConfig *rest.Config
         fromConfig *rest.Config
}
```
1. Is  this somewhat you mean here - `Then at the cmd/syncer/cmd level you will use the option structs from the options package and turn e.g. the kubeconfig path name into rest.Configs`? And please if you can elaborate if I'm not on the same package?
2. If so, Is ```SyncerOptions``` a correct naming convention?

Comment (RAMESSESII2) [reply cmd/syncer/cmd/syncer.go:114]: Hi @sttts, I agree, it makes sense to put Options struct in its own package. And then let's say I create new Struct which will eventually be composed of options and the derived variables(toConfig and fromConfig). Something like this -
```
type SyncerOptions struct{
         options *options.Options
         toConfig *rest.Config
         fromConfig *rest.Config
}
```
1. Is this somewhat you mean here - Then at the cmd/syncer/cmd level you will use the option structs from the options package and turn e.g. the kubeconfig path name into rest.Configs? And please if you can elaborate if I'm not on the same package?
2. If so, Is SyncerOptions a correct naming convention?

**Comment by sttts** [comment]
@RAMESSESII2 please see https://github.com/RAMESSESII2/kcp/pull/2.

**Comment by sttts** [comment]
As background:

Syncer is simple enough that we can create the config just in the `Run` func. In bigger setups (like kube-apiserver) we would have an intermediate Config or Server struct holding the config. But here that would be overkill.

The structure in the options package is always the same, like in https://github.com/RAMESSESII2/kcp/pull/2. It's a pattern that proved to work well for composition of components in one binary.

**Comment by RAMESSESII2** [comment]
> Syncer is simple enough that we can create the config just in the `Run` func. In bigger setups (like kube-apiserver) we would have an intermediate Config or Server struct holding the config. But here that would be overkill.
Ah, sure, makes sense and primarily that eliminates the problem of keeping track of those those two vaibales that are essentially derived from Options itself. Thanks 😄

**Comment by sttts** [comment]
I made the linter unhappy:
```
--- a/cmd/syncer/options/options.go
+++ b/cmd/syncer/options/options.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 
 	"github.com/spf13/pflag"
+
 	"k8s.io/component-base/logs"
 
 	nscontroller "github.com/kcp-dev/kcp/pkg/reconciler/namespace"
```
Can you fix that? Then it's ready to go.

**Comment by RAMESSESII2** [comment]
> I made the linter unhappy:
> Can you fix that? Then it's ready to go.

yupp, IG we're good now 😄

**Review by sttts — APPROVED** [review]

**Review by ncdc — APPROVED** [review]
